File: C/Algorithms/985.c
#include <stdlib.h>

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int *sumEvenAfterQueries(int *A, int ASize, int **queries, int queriesSize,
                         int *queriesColSize, int *returnSize) {
  int evenSum = 0, v;
  for (int i = 0; i < ASize; i++) {
    v = *(A + i);
    if (v % 2 == 0)
      evenSum += v;
  }
  int *result = malloc(queriesSize * sizeof(*result)), r_idx = 0, idx,
      old_value, new_value;
  for (int i = 0; i < queriesSize; i++) {
    idx = queries[i][1];
    old_value = *(A + idx);
    new_value = old_value + queries[i][0];

    // add to evenSum if even:
    if (new_value % 2 == 0)
      evenSum += new_value;
    // subtract if old value was even
    if (old_value % 2 == 0)
      evenSum -= old_value;

    // Update A[idx], set to result.
    *(A + idx) = new_value;
    *(result + r_idx++) = evenSum;
  }
  *returnSize = queriesSize;
  return result;
}

File: Rust/Algorithms/985.rs
impl Solution {
    pub fn sum_even_after_queries(
        mut a: Vec<i32>,
        queries: Vec<Vec<i32>>,
    ) -> Vec<i32> {
        let mut evenSum: i32 = a.iter().filter(|x| *x % 2 == 0).sum();
        let mut result = Vec::with_capacity(queries.len());

        for query in queries.into_iter() {
            let (value, index) = (query[0], query[1] as usize);
            let old_value = a[index];
            let new_value = value + old_value;

            // Add new value if its even:
            if new_value % 2 == 0 {
                evenSum += new_value;
            }
            // Subtract old value if it was even.
            if old_value % 2 == 0 {
                evenSum -= old_value;
            }
            a[index] = new_value;
            result.push(evenSum);
        }
        result
    }
}
